### PR TITLE
Do not bother optimizing impossible functions.

### PR DESCRIPTION
**Pull request by cjgillot**
This is currently checked by `ConstProp`, but I see no reason to restrict it to ConstProp only.

### CODE REVIEW

File: compiler/rustc_mir_transform/src/lib.rs [hunk 1]
```diff
@@ -481,6 +482,54 @@ fn run_optimization_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
...
+    // Check if it's even possible to satisfy the 'where' clauses
+    // for this item.
+    // This branch will never be taken for any normal function.
+    // However, it's possible to `#!feature(trivial_bounds)]` to write
+    // a function with impossible to satisfy clauses, e.g.:
+    // `fn foo() where String: Copy {}`
+    //
+    // We don't usually need to worry about this kind of case,
+    // since we would get a compilation error if the user tried
+    // to call it. However, since we can do const propagation
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/lib.rs:494]: this still mentions const prop explicitly. Should probably just talk about opts.

**Review by oli-obk — COMMENTED** [review]
Should we just replace the body with an empty one while we're at it? Or change mir building to just generate an empty body to begin with?

**Comment by cjgillot** [comment]
For MIR building, I don't think so. We still want borrowck and other analyses to run.
I'll implement clearing the body of statements.

**Review by JakobDegen — APPROVED** [review]

**Comment by wesleywiser** [comment]
This should have a positive impact on compiler performance right? Either way, seems like a good idea to:

@bors rollup=never

**Comment by pnkfelix** [comment]
> This should have a positive impact on compiler performance right?

* Seems like it didn't have the expected positive impact
* primary regressions are to libc; most of secondary are to variants many-assoc-items (plus some check incr-unchanged on three other secondaries).
* in any case, the measured impact looks negligible to me.

@rustbot label: +perf-regression-triaged

**Pull request merged**